ClassNotFoundException: com.vmware.vim25.PerfCounterInfoInt

Hi, I use yavijava to connect vcenter 6.0.0 , When I invoke getPerfCounter method in PerformanceManager to get Performance metric value it gets the following error:

`
java.lang.ClassNotFoundException: com.vmware.vim25.PerfCounterInfoInt
    at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:425)
    at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:358)
    at java.lang.Class.forName0(Native Method)
    at java.lang.Class.forName(Class.java:191)
    at org.doublecloud.ws.util.TypeUtil.getVimClass(TypeUtil.java:100)
    at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:322)
    at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:329)
    at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:322)
    at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:204)
    at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:120)
    at com.vmware.vim25.ws.SoapClient.unMarshall(SoapClient.java:212)
    at com.vmware.vim25.ws.WSClient.invoke(WSClient.java:93)
    at com.vmware.vim25.ws.VimStub.retrieveProperties(VimStub.java:106)
    at com.vmware.vim25.mo.PropertyCollector.retrieveProperties(PropertyCollector.java:98)
    at com.vmware.vim25.mo.ManagedObject.retrieveObjectProperties(ManagedObject.java:146)
    at com.vmware.vim25.mo.ManagedObject.getCurrentProperty(ManagedObject.java:170)
    at com.vmware.vim25.mo.PerformanceManager.getPerfCounter(PerformanceManager.java:59)
2016-03-17 15:12:08.724 ERROR [WSClient.invoke(96)] Exception caught while invoking method: RetrieveProperties
java.rmi.RemoteException: Exception in SoapClient.invoke:; nested exception is: 
    java.lang.NullPointerException
    at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:123)
    at com.vmware.vim25.ws.SoapClient.unMarshall(SoapClient.java:212)
    at com.vmware.vim25.ws.WSClient.invoke(WSClient.java:93)
    at com.vmware.vim25.ws.VimStub.retrieveProperties(VimStub.java:106)
    at com.vmware.vim25.mo.PropertyCollector.retrieveProperties(PropertyCollector.java:98)
    at com.vmware.vim25.mo.ManagedObject.retrieveObjectProperties(ManagedObject.java:146)
    at com.vmware.vim25.mo.ManagedObject.getCurrentProperty(ManagedObject.java:170)
    at com.vmware.vim25.mo.PerformanceManager.getPerfCounter(PerformanceManager.java:59)
    at com.oneapm.si.rm.plugin.vimapi.vcenter.VCenterPerformanceStatement.filterCounterInfo(VCenterPerformanceStatement.java:206)
    at com.oneapm.si.rm.plugin.vimapi.vcenter.VCenterPerformanceStatement.doRealTime(VCenterPerformanceStatement.java:114)
    at com.oneapm.si.rm.plugin.vimapi.vcenter.VCenterPerformanceStatement.execute(VCenterPerformanceStatement.java:92)
    at com.oneapm.si.rm.plugin.vimapi.vcenter.VCenterPerformanceStatementTest.test(VCenterPerformanceStatementTest.java:52)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:44)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:15)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:41)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:20)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:28)
    at org.junit.runners.BlockJUnit4ClassRunner.runNotIgnored(BlockJUnit4ClassRunner.java:79)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:71)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:49)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:193)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:52)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:191)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:42)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:184)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:236)
    at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:50)
    at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:467)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:683)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:390)
    at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:197)
Caused by: java.lang.NullPointerException
    at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:219)
    at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:322)
    at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:329)
    at com.vmware.vim25.ws.XmlGenDom.fromXml(XmlGenDom.java:322)
    at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:204)
    at com.vmware.vim25.ws.XmlGenDom.fromXML(XmlGenDom.java:120)
`

Java version is: 1.7.0_79
vSphere is: 6.0.0
yavijava is: 6.0.03

--- a/src/main/java/com/vmware/vim25/ws/XmlGenDom.java
+++ b/src/main/java/com/vmware/vim25/ws/XmlGenDom.java
@@ -241,7 +241,12 @@ class XmlGenDom extends XmlGen {
             for (int i = 0; i < subNodes.size(); i++) {
                 Element e = subNodes.get(i);
                 String xsiType = e.attributeValue(SoapConsts.XSI_TYPE);
-                Object o = fromXml(TypeUtil.getVimClass(xsiType == null ? arrayItemTypeName : xsiType), subNodes.get(i));
+                Class<?> itemClass = TypeUtil.getVimClass(xsiType == null ? arrayItemTypeName : xsiType);
+                if (itemClass == null) {
+                    log.debug("Unknown xsi:type '{}' for array element, using base type {}", xsiType, arrayItemTypeName);
+                    itemClass = clazz;
+                }
+                Object o = fromXml(itemClass, subNodes.get(i));
                 Array.set(ao, i, o);
             }
             return ao;
@@ -293,8 +298,25 @@ class XmlGenDom extends XmlGen {
 
             Class fRealType = fType;
             String xsiType = e.attributeValue(SoapConsts.XSI_TYPE);
-            if (xsiType != null && (!xsiType.startsWith("xsd:"))) {
-                fRealType = TypeUtil.getVimClass(xsiType);
+            if (xsiType != null && xsiType.startsWith("ArrayOf")) {
+                String itemTypeName = xsiType.substring("ArrayOf".length());
+                Class<?> itemClass = TypeUtil.getVimClass(itemTypeName);
+                if (itemClass != null) {
+                    field.set(obj, fromXML(itemTypeName + "[]", e));
+                }
+                else {
+                    log.debug("Unknown ArrayOf item type '{}', skipping field", itemTypeName);
+                }
+                continue;
+            }
+            else if (xsiType != null && (!xsiType.startsWith("xsd:"))) {
+                Class resolvedType = TypeUtil.getVimClass(xsiType);
+                if (resolvedType != null) {
+                    fRealType = resolvedType;
+                }
+                else {
+                    log.debug("Unknown xsi:type '{}', falling back to declared field type {}", xsiType, fType.getSimpleName());
+                }
             }
 
             if (fRealType == ManagedObjectReference.class) { // MOR

--- a/src/test/java/com/vmware/vim25/ws/XmlGenDomTest.java
+++ b/src/test/java/com/vmware/vim25/ws/XmlGenDomTest.java
@@ -192,4 +192,22 @@ public class XmlGenDomTest {
             (com.vmware.vim25.NoPermission) mp.getFault().getFault();
         Assert.assertEquals("System.Read", np.privilegeId);
     }
+
+    @Test
+    public void testFromXML_PerfCounterInfoWithUnknownSubtype_ParsesGracefully() throws Exception {
+        InputStream inputStream = new FileInputStream(
+            new File("src/test/java/com/vmware/vim25/ws/xml/PerfCounterInfoWithUnknownSubtype.xml"));
+        XmlGenDom xmlGenDom = new XmlGenDom();
+        ObjectContent objectContent = (ObjectContent) xmlGenDom.fromXML("ObjectContent", inputStream);
+
+        Assert.assertNotNull(objectContent.getPropSet());
+        Assert.assertEquals(1, objectContent.getPropSet().length);
+
+        PerfCounterInfo[] counters = (PerfCounterInfo[]) objectContent.getPropSet()[0].getVal();
+        Assert.assertNotNull("counter array should not be null", counters);
+        Assert.assertEquals("should have parsed both counters", 2, counters.length);
+
+        Assert.assertEquals("first counter key", 1, counters[0].getKey());
+        Assert.assertEquals("unknown-subtype counter key should be 369", 369, counters[1].getKey());
+    }
 }

--- a/src/test/java/com/vmware/vim25/ws/xml/PerfCounterInfoWithUnknownSubtype.xml
+++ b/src/test/java/com/vmware/vim25/ws/xml/PerfCounterInfoWithUnknownSubtype.xml
@@ -1,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"
+                  xmlns:xsd="http://www.w3.org/2001/XMLSchema"
+                  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+    <soapenv:Body>
+        <RetrievePropertiesResponse xmlns="urn:vim25">
+            <returnval>
+                <obj type="PerformanceManager">PerfMgr</obj>
+                <propSet>
+                    <name>perfCounter</name>
+                    <val xsi:type="ArrayOfPerfCounterInfo">
+                        <PerfCounterInfo xsi:type="PerfCounterInfo">
+                            <key>1</key>
+                            <nameInfo><label>Usage</label><summary>CPU usage</summary><key>usage</key></nameInfo>
+                            <groupInfo><label>CPU</label><summary>CPU</summary><key>cpu</key></groupInfo>
+                            <unitInfo><label>%</label><summary>Percentage</summary><key>percent</key></unitInfo>
+                            <rollupType>none</rollupType>
+                            <statsType>rate</statsType>
+                            <level>4</level>
+                            <perDeviceLevel>4</perDeviceLevel>
+                        </PerfCounterInfo>
+                        <PerfCounterInfo xsi:type="PerfCounterInfoInt">
+                            <key>369</key>
+                            <nameInfo><label>VXLAN tx throughput</label><summary>Transmitted packet rate</summary><key>throughput.vds.txTotal</key></nameInfo>
+                            <groupInfo><label>Network</label><summary>Network</summary><key>net</key></groupInfo>
+                            <unitInfo><label>num</label><summary>Number</summary><key>number</key></unitInfo>
+                            <rollupType>summation</rollupType>
+                            <statsType>delta</statsType>
+                            <level>4</level>
+                            <perDeviceLevel>4</perDeviceLevel>
+                            <enabledByDefault>false</enabledByDefault>
+                            <enabled>false</enabled>
+                        </PerfCounterInfo>
+                    </val>
+                </propSet>
+            </returnval>
+        </RetrievePropertiesResponse>
+    </soapenv:Body>
+</soapenv:Envelope>

